File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
using System;

class RemoteControlCar
{
    private int speed;
    private int batteryDrain;
    private int batteryCharge;
    private int distanceDriven;
    
    public RemoteControlCar(int speed, int batteryDrain)
    {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
        this.batteryCharge = 100;
        this.distanceDriven = 0;
    }

    public bool BatteryDrained() => batteryCharge < batteryDrain;

    public int DistanceDriven() => distanceDriven;

    public void Drive()
    {
        if (!BatteryDrained())
        {
            distanceDriven += speed;
            batteryCharge -= batteryDrain;
        }
    }

    public static RemoteControlCar Nitro() => new RemoteControlCar(50, 4);
}

class RaceTrack
{
    private int distance;

    public RaceTrack(int distance) => this.distance = distance;

    public bool TryFinishTrack(RemoteControlCar car)
    {
        while (car.DistanceDriven() < distance)
        {
            if (car.BatteryDrained())
                return false;
            car.Drive();
        }
        return true;
    }
}

File: solutions/java/need-for-speed/2/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    int speed;
    int batteryDrainRate;
    int batteryCharge = 100;
    int distanceDriven = 0;
    
    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
	this.batteryDrainRate = batteryDrain;
    }

    public boolean batteryDrained() {
        return batteryCharge < batteryDrainRate;
    }

    public int distanceDriven() {
        return distanceDriven;
    }

    public void drive() {
        if (!batteryDrained()) {
	    distanceDriven += speed;
	}
	batteryCharge = Math.max(0, batteryCharge - batteryDrainRate);
    }

    public static NeedForSpeed nitro() {
	return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {

    int distance;
    
    RaceTrack(int distance) {
	this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        return (double)distance / car.speed * car.batteryDrainRate <= 100.0;
    }
}
